### DISCUSSION TITLE
continuous variable child with discrete variable parent

### DISCUSSION THREAD

**Discussion by borjasor** [pgmpy/pgmpy#1752] [Q&A]
First of all congratulate the team for 'pgmpy', it is a great tool!

Hopefully someone can guide me on how to implement this case.
I define a model with two nodes:

model = BayesianNetwork([('A', 'B')])

'A' is a discrete variable with 4 estates, its CPD is defined by:
cpd_A = TabularCPD(variable='A', variable_card=4, values=[[0.25], [0.25], [0.25], [0.25]], state_names={'A': ['one', 'two', 'three', 'four']})

'B' is a continuous variable with a different 'PDF' for each evidence of 'A'; 
one_factor = ContinuousFactor(['B'], pdf_one)
two_factor = ContinuousFactor(['B'], pdf_two)
three_factor = ContinuousFactor(['B'], pdf_three)
four_factor = ContinuousFactor(['B'], pdf_four)

The question is how can I assign these PDFs of B to the BayesianNetwork 'model'?

Thank you for your help.

**Comment by ankurankan** [comment]
@borjasor Unfortunately, continuous and hybrid (continuous + discrete) factors are not fully supported yet.

**Reply by borjasor** [reply]
Thank you @ankurankan , I assume the only way to proceed with pgmpy (in my case) would be to discretize the continuous variable. However, there seems to be quite a few issues when trying to do this:
1) The examples in 'discretize.py' seems to refer to a nonexisting class 'ContinuousNode'
2) Assuming that 'ContinuousNode' = ContinuousFactor, when using 'RoundingDiscretizer' or 'UnbiasedDiscretizer' there are issues as both make use of 'cdf' = cumulative density function which there does not seem to be way to define in 'ContinuousFactor'.
3) When using class 'BaseDiscretizer', currently is defined in a way that uses 'ABCMeta' class incorrectly as it is a 'meta class', not a 'base class', so you will need to change the class definition to:
class BaseDiscretizer(**metaclass**=ABCMeta):

Even when fixing the above, when using 'discretize' on 'BaseDiscretizer' we get an error: 
TypeError: Can't instantiate abstract class BaseDiscretizer with abstract method get_discrete_values

**Reply by ankurankan** [reply]
@borjasor Yes, the discretizer methods need to be fixed. The continuous variable implementation (still incomplete) has changed quite a bit since the methods were written. I would just avoid using that for now. I just added a new function pgmpy.utils.discretize (https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/utils/utils.py#L119), which is a simple wrapper around pandas.cut methods that you can use for discretization.

**Reply by borjasor** [reply]
Thank you again @ankurankan , I will look into this.

**Reply by borjasor** [reply]
@ankurankan after noticing recent additions to pgmpy: [FunctionalCPD](https://github.com/pgmpy/pgmpy/commit/82c7cdd1d55ff82ae340d63896594461f31257d0#diff-510efa4f418a01bfa5fe3b6f37b7b27f69f442194cfcb29b3e544b35523c311b) and [FunctionalBayesianNetwork](https://github.com/pgmpy/pgmpy/commit/82c7cdd1d55ff82ae340d63896594461f31257d0#diff-28ab977cba7c2706ffa0d04f922e4c1b2795cbccaaf770ee7d9a4e1e129f3edc)

I thought important to revisit this original question. Does this mean that continuous and hybrid factors are now fully supported under FunctionalCPD? Also the only way to use these factors is within a FunctionalBayesianNetwork or can they be used also with a BayesianNetwork model also including Tabular CPDs? Finally, how does this hybrid support affect Parameter Learning as this seems to only be supported for [Discrete variables](https://pgmpy.org/examples/Learning%20Parameters%20in%20Discrete%20Bayesian%20Networks.html)?

**Reply by ankurankan** [reply]
@borjasor The aim with Functional CPD is to allow users to specify any arbitrary distribution including mixtures of discrete, ordinal, and continuous variables. It is still a work in progress, so the functionality is limited to data simulation. The current plan is to keep Bayesian Network (maybe rename it to discrete BN), LinearGaussianBayesianNetwork, and Functional Bayesian Network separate as they all don't share a common feature set yet. The user would need to choose the BN class depending on their data. In future, we can combine the classes together once we have enough features to have a common API for them.

> Finally, how does this hybrid support affect Parameter Learning as this seems to only be supported for [Discrete variables](https://pgmpy.org/examples/Learning%20Parameters%20in%20Discrete%20Bayesian%20Networks.html)?

It's not supported yet but @Nuna7 is working on it should be in the dev branch soon: https://github.com/pgmpy/pgmpy/pull/1893
